### ISSUE TITLE
Writing events to a stream is retried even though POST request has already returned 500

### ISSUE THREAD

**Issue by razvanz** [EventStore/EventStore#1358]
`Event Store 3.9.3.0` - 3-node cluster

In my use case, I'm writing events by POST-ing to a stream, to which EventStore sometimes returns 500. Nevertheless, it seems that the events for which error was returned, are written to the stream, which, for me, is unexpected behavior. An error response should mean the transaction has failed and I should be able to retry it without expecting any traces from the previous operation.

```
[PID:25040:033 2017.07.12 14:08:46.619 INFO  CoreProjectionCheckp] Failed to write events to stream $ce-Invoice. Error: CommitTimeout
[PID:25040:033 2017.07.12 14:08:46.619 INFO  CoreProjectionCheckp] Retrying write to $ce-Invoice
[PID:25040:026 2017.07.12 14:08:46.620 TRACE TwoPhaseRequestManag] IDEMPOTENT WRITE TO STREAM ClientCorrelationID f8426d5a-ef4e-44a5-9c95-f1cfe95e5944, EventStreamId: $ce-Invoice, CorrelationId: f8426d5a-ef4e-44a5-9c95-f1cfe95e5944, FirstEventNumber: 2700192, LastEventNumber: 2700192.
```

**Comment by gregoryyoung** [comment]
Are there errors in the log? The log entries here before the idempotent write are from projections which are most likely unrelated

**Comment by razvanz** [comment]
Yes, those are errors from the log. 

I think we are running into #803, since I can see lots of `System.ArgumentException: complete TF position required` in the error log and we are using `fromAll` projections but don't it that could be related. I would expect projections to be triggered after the write has successfully completed.

**Comment by gregoryyoung** [comment]
That is most likely unrelated. What is the full http response of the 500?

**Comment by razvanz** [comment]
The response has an empty body.

**Comment by razvanz** [comment]
I've upgraded the cluster to `4.0.1` today and it seems that the error persists. Let me know if you want me to provide logs.

**Comment by pgermishuys** [comment]
Hi @razvanz 
1. A reproduction would be very helpful.
2. You say you are running into #803? What exactly are you observing?

**Comment by razvanz** [comment]
1. I will try to see if I can reproduce it in a minimal environment
2. Before upgrading to `4.0.1` I could see `System.ArgumentException: complete TF position required` in the logs. Now I cannot see them anymore, however, the 500s persists.

**Comment by razvanz** [comment]
I tried reproducing the error locally but wasn't able to. However, I turned on the `LogHttpRequests` option and it seems writing the events times out:

```
[PID:02688:027 2017.07.17 10:35:24.017 DEBUG HttpEntityManager   ] HTTP Response                                                                                                         
07/17/2017 10:35:24                                                                                                                                                                      
500 Write timeout                                                                                                                                                                        
host: eventstore.db.vpc:2113                                                                                                                                                             
Accept-Encoding: gzip, deflate                                                                                                                                                           
Content-Type: application/vnd.eventstore.events+json                                                                                                                                     
user-agent: billy-api/1                                                                                                                                                                  
X-Forwarded-For: 172.30.22.134                                                                                                                                                           
X-Forwarded-Port: 2113                                                                                                                                                                   
X-Forwarded-Proto: http                                                                                                                                                                  
Content-Length: 3982                                                                                                                                                                     
Connection: keep-alive 
```

`CommitTimeoutMs` is configured at `4000` and there aren't that many write requests hitting the cluster. Any ideas?

**Comment by pgermishuys** [comment]
Writes timing out in a cluster can be due to any number of reasons.

For example:
- Acknlowedgement from the other nodes in the cluster are taking awhile to get back to the master and in that case the write can timeout which could be attributed to 
   - Slow network
   - Slow disk

Without logs, it's difficult to say.

**Comment by razvanz** [comment]
What logs would help identify the issue?

**Comment by pgermishuys** [comment]
The Event Store logs from each of the nodes in the cluster.

**Comment by razvanz** [comment]
I would like not to make them public. Can you give me an email address to send them to?

**Comment by pgermishuys** [comment]
ofcourse. pieter at geteventstore dot com

**Comment by Faq** [comment]
@razvanz problem still exists? Did you tried with v4.0.3?

**Comment by razvanz** [comment]
@Faq Forgot to update this. 

It was fixed with `v4.0.1.4` so I assume everything released after that should not have the issue.
